### PR TITLE
Security for Elastic Stack in Docker implemented

### PR DESCRIPTION
**Pull request by SitoRBJ**
Hello team, 

This PR solves #186 

Best regards,

Alfonso Ruiz-Bravo

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -37,6 +42,20 @@ services:
       - network.host=0.0.0.0
       - bootstrap.memory_lock=true
       - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
+      - ELASTICSEARCH_URL=https://elasticsearch:9200
+      - ELASTICSEARCH_IP=https://elasticsearch
+      - ELASTICSEARCH_PORT=9200
+      - SETUP_PASSWORDS=yes
```
Comment (jesuslinares) [inline docker-compose.yml]: A better name would be something like:
`XPACK_SECURITY_ENABLED`

Comment (jesuslinares) [reply docker-compose.yml]: Every related variable should start with `XPACK_SECURITY_`.
Example:
`XPACK_SECURITY_ELASTIC_PASSWORD`

File: docker-compose.yml [hunk 2]
```diff
@@ -37,6 +42,20 @@ services:
       - network.host=0.0.0.0
       - bootstrap.memory_lock=true
       - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
+      - ELASTICSEARCH_URL=https://elasticsearch:9200
```
Comment (jesuslinares) [inline docker-compose.yml]: I think it should be:
`ELASTICSEARCH_PROTOCOL="https"`
`ELASTICSEARCH_IP="elasticsearch"`
`ELASTICSEARCH_PORT="9200"`

Then, you can build a variable (not as input):
`ELASTICSEARCH_URL="$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_IP"`

File: docker-compose.yml [hunk 3]
```diff
@@ -37,6 +42,20 @@ services:
...
+      - SETUP_PASSWORDS=yes
+      - ADMIN_USER=admin_user
+      - ADMIN_PASS=admin_pass
+      - ELASTIC_PASSWORD=elastic_pass
+      - KIBANA_PASS=kibana_pass
+      - LOGSTASH_PASS=logstash_pass
+      - APM_SYSTEM_PASS=apmSystem_pass
+      - BEATS_SYSTEM_PASS=beatsSystem_pass
+      - REMOTE_USER_PASS=remoteUser_pass
+      - CA_PASS=ca_pass
```
Comment (jesuslinares) [inline docker-compose.yml]: Change to `XPACK_SECURITY_CA_PASSPHRASE`.

File: docker-compose.yml [hunk 4]
```diff
@@ -37,6 +42,20 @@ services:
...
+      - CERT_DNS=elasticsearch
```
Comment (jesuslinares) [inline docker-compose.yml]: Change to `XPACK_SECURITY_CERTIFICATE_DNS`.

File: elasticsearch/Dockerfile [hunk 5]
```diff
@@ -29,8 +29,13 @@ ENV ELASTIC_CLUSTER="false" \
     CLUSTER_MAX_NODES="1" \
     CLUSTER_DELAYED_TIMEOUT="1m"
 
+
+
 ADD https://raw.githubusercontent.com/wazuh/wazuh/$TEMPLATE_VERSION/extensions/elasticsearch/6.x/wazuh-template.json /usr/share/elasticsearch/config
 
+# CA cert for Transport SSL
+ADD https://github.com/SitoRBJ/test-ca/raw/master/elastic-CA.zip /usr/share/elasticsearch/config
```
Comment (jesuslinares) [inline elasticsearch/Dockerfile]: Add a test CA/certs in the repository.

File: elasticsearch/config/config_secure.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  echo "
+instances:
+- name: \"elasticsearch\"
+  dns: 
+    - $CERT_DNS
+" > instances.yml
+
+  unzip elastic-CA.zip
+  /usr/share/elasticsearch/bin/elasticsearch-certutil cert --pem -in instances.yml --out certs.zip --ca-cert server.CA-signed.pem --ca-key server.CA.key  --ca-pass $CA_PASS
+  unzip certs.zip
```
Comment (jesuslinares) [inline elasticsearch/config/config_secure.sh]: Remove zip files.

File: elasticsearch/config/config_secure.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  popd
+
+  chown elasticsearch: /usr/share/elasticsearch/config/server.CA-signed.pem
+  chown -R elasticsearch: /usr/share/elasticsearch/config/elasticsearch
+  chmod 774 /usr/share/elasticsearch/config/server.CA-signed.pem
```
Comment (jesuslinares) [inline elasticsearch/config/config_secure.sh]: Permissions should be 440 for files.

File: kibana/Dockerfile [hunk 8]
```diff
@@ -75,4 +80,16 @@ RUN ./welcome_wazuh.sh
 
 RUN /usr/local/bin/kibana-docker --optimize
 
+
+# Set Wazuh custom login image
```
Comment (jesuslinares) [inline kibana/Dockerfile]: Great welcome screen, but we are not going to customize it yet.

File: docker-compose.yml (hunk 4 above)
Comment (SitoRBJ) [reply docker-compose.yml]: Changed

File: docker-compose.yml (hunk 1 above)
Comment (SitoRBJ) [reply docker-compose.yml]: Changed

File: elasticsearch/config/config_secure.sh (hunk 6 above)
Comment (SitoRBJ) [reply elasticsearch/config/config_secure.sh]: Removed

File: kibana/Dockerfile (hunk 8 above)
Comment (SitoRBJ) [reply kibana/Dockerfile]: Moved

File: elasticsearch/config/config_secure.sh (hunk 7 above)
Comment (SitoRBJ) [reply elasticsearch/config/config_secure.sh]: Changed, we will test with 440 permissions.

File: docker-compose.yml (hunk 2 above)
Comment (SitoRBJ) [reply docker-compose.yml]: Changed

File: elasticsearch/Dockerfile (hunk 5 above)
Comment (SitoRBJ) [reply elasticsearch/Dockerfile]: Changed

File: create_CA.sh [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    openssl req -verbose -new -key ${KEY_CA} -out ${CSR_CA} -sha256
+    echo "Self sign certificate"
+    openssl ca -extensions v3_ca -out ${SIGNED_CA} -keyfile ${KEY_CA} -verbose -selfsign -md sha256 -days ${DAYS} -infiles ${CSR_CA}
+    echo "Self signed CA created"
+    cp ${KEY_CA} elasticsearch/config/
+    cp ${SIGNED_CA} elasticsearch/config/
+    cp ${SIGNED_CA} kibana/config/
+    cp ${SIGNED_CA} logstash/config/
+    rm ${SIGNED_CA} ${KEY_CA} ${CSR_CA}
+    sed -i 's:ARG SECURITY_CA_PEM_LOCATION="config/server.TEST-CA-signed.pem":ARG SECURITY_CA_PEM_LOCATION="config/'$SIGNED_CA'":g' elasticsearch/Dockerfile
```
Comment (jesuslinares) [inline create_CA.sh]: Remove `sed` commands. The user responsibility is to edit the docker compose according to the certificate names.

File: create_CA.sh [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    create_own_CA
+}
+
+
+creation_type ()
+{
+    if [[ $ACTION == "create_CA" ]]; then
+        create_own_CA 
+    elif [[ $ACTION == "full_creation" ]]; then
+        full_own_creation
```
Comment (jesuslinares) [inline create_CA.sh]: Is it possible to remove this function?. Can we use a specific path?

**Pull request closed without merging**